detox/src/artifacts/core/artifact/RecordingArtifact.js

```diff
@@ -0,0 +1,96 @@
+const DetoxRuntimeError = require('../../../errors/DetoxRuntimeError');
+
+class RecordingArtifact {
+  constructor() {
+    this._startPromise = null;
+    this._stopPromise = null;
+    this._savePromise = null;
+    this._discardPromise = null;
+  }
+
+  start() {
+    this._assertRecordingIsNotBeingResumed();
+
+    if (!this._startPromise) {
+      this._startPromise = this.doStart();
+    }
+
+    return this._startPromise;
+  }
+
+  stop() {
+    this._assertRecordingHasBeenStarted();
+
+    if (!this._stopPromise) {
+      this._stopPromise = this.doStop();
+    }
+
+    return this._stopPromise;
+  }
+
+  save(artifactPath) {
+    this._assertRecordingIsNotBeingDiscarded();
+
+    if (!this._savePromise) {
+      this._savePromise = this.stop().then(() => this.doSave(artifactPath));
+    }
+
+    return this._savePromise;
+  }
+
+  discard() {
+    this._assertRecordingIsNotBeingSaved();
+
+    if (!this._discardPromise) {
+      this._discardPromise = this.stop().then(() => this.doDiscard());
+    }
+
+    return this._discardPromise;
+  }
+
+  async doStart() {}
+
+  async doStop() {}
+
+  async doSave(artifactPath) {}
+
+  async doDiscard() {}
+
+  _assertRecordingIsNotBeingResumed() {
```

Comment:
You’re right. Now it is not user-facing, but honestly I was thinking that we wanted to make it available in next versions.

I think we indeed can remove most of the checks, but on other hand they served me a great debugging value whilst writing lifecycles. If anybody goes to change something there or adds a new lifecycle, it can help while testing.

Secondly, sometimes we might need a smarter logic inside. E.g., if on `SIGINT` we have to clean up allocated temporary artifact files, it is easier to call `.discard()` from ArtifactsManager for each artifact rather than implement status checks to see if we need to `.stop()` it before `.discard()`-ing it. Smarter implementation of `.discard()` gives less headache when you just need to `.discard()` something.

Reply:
I though that lifecycle is controlled by ArtifactManager, its a one time implementation, am I wrong ?

Reply:
It is, but indirectly.
I prefer to keep lifecycles of a `SnapshotArtifact` and a `RecordingArtifact` as separate entities.
They are different enough in my opinion to not try to stitch them together.
In any case, `ArtifactsManager` is responsible for overall hooks (`onStart`, `onExit`, ...) execution and it has a couple of global (cross-lifecycle) handlers:
* It keeps registering artifacts, so it can discard them in an emergency.
* It provides `enqueueFinalizationTask` method to lifecycles - to coordinate finalization, e.g. to make it less CPU-intensive.

In other words, `ArtifactsManager` alone has enough of a headache to handle, so it delegates minutiae to "smaller" and more specific lifecycles per artifact type.